File: examples/javascript/doc-du-lieu.js
#!/usr/bin/env node
// Doc va tra cuu bo du lieu 12 Dia Chi bang JavaScript thuan.
//
// Chay:  node doc-du-lieu.js
// Khong can thu vien ngoai.

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const GOC = join(dirname(fileURLToPath(import.meta.url)), "..", "..", "data");

// Bo doc CSV toi thieu nhung dung RFC 4180: o duoc boc trong dau nhay kep
// co the chua dau phay, va hai dau nhay lien nhau la mot dau nhay that.
function docCSV(vanBan) {
  const dong = [];
  let o = [], cur = "", trongNhay = false;
  for (let i = 0; i < vanBan.length; i++) {
    const c = vanBan[i];
    if (trongNhay) {
      if (c === '"' && vanBan[i + 1] === '"') { cur += '"'; i++; }
      else if (c === '"') trongNhay = false;
      else cur += c;
    } else if (c === '"') trongNhay = true;
    else if (c === ",") { o.push(cur); cur = ""; }
    else if (c === "\n") { o.push(cur); dong.push(o); o = []; cur = ""; }
    else if (c !== "\r") cur += c;
  }
  if (cur || o.length) { o.push(cur); dong.push(o); }
  const [ten, ...than] = dong.filter((d) => d.some((x) => x !== ""));
  return than.map((d) => Object.fromEntries(ten.map((t, i) => [t, d[i] ?? ""])));
}

const chi = docCSV(readFileSync(join(GOC, "12-dia-chi.csv"), "utf8"));
console.log(`Nap ${chi.length} dong tu 12-dia-chi.csv`);

for (const c of chi) {
  console.log(`  ${c.chi.padEnd(5)} ${c.con_giap.padEnd(6)} ${c.ngu_hanh.padEnd(5)} ${c.khung_gio}`);
}

// Cot ghi_chu la noi ghi diem cac truong phai bat dong.
const coGhiChu = chi.filter((c) => c.ghi_chu.trim());
console.log(`\nDong co ghi chu bat dong: ${coGhiChu.length}`);
for (const c of coGhiChu) console.log(`  ${c.chi}: ${c.ghi_chu}`);

File: examples/python/doc-du-lieu.py
#!/usr/bin/env python3
"""Doc va tra cuu bo du lieu 60 Hoa Giap.

Chay:  python3 doc-du-lieu.py
Khong can thu vien ngoai.
"""
import csv
import pathlib

GOC = pathlib.Path(__file__).resolve().parents[2] / "data"


def nap(slug):
    with open(GOC / f"{slug}.csv", encoding="utf-8", newline="") as f:
        return list(csv.DictReader(f))


def main():
    hoa_giap = nap("60-hoa-giap")
    print(f"Nap {len(hoa_giap)} dong tu 60-hoa-giap.csv")

    # Tra can chi cua mot nam: cot cac_nam liet ke cac nam ngan bang dau cham phay.
    nam = "1984"
    for h in hoa_giap:
        if nam in [x.strip() for x in h["cac_nam"].split(";")]:
            print(f"Nam {nam}: {h['can_chi']}, nap am {h['ten_nap_am']} ({h['ngu_hanh_nap_am']})")
            break

    # Dem theo ngu hanh nap am.
    dem = {}
    for h in hoa_giap:
        dem[h["ngu_hanh_nap_am"]] = dem.get(h["ngu_hanh_nap_am"], 0) + 1
    print("So cap theo ngu hanh nap am:")
    for hanh, n in sorted(dem.items(), key=lambda x: -x[1]):
        print(f"  {hanh}: {n}")

    # Cot ghi_chu la noi ghi diem cac truong phai bat dong.
    co_ghi_chu = [h for h in hoa_giap if h["ghi_chu"].strip()]
    print(f"So dong co ghi chu bat dong: {len(co_ghi_chu)}")


if __name__ == "__main__":
    main()
